### PR TITLE
Store snapshot statistics & print snapshot size

### LINKED ISSUE #693: Print out the backup size when listing snapshots (enhancement)
## Output of `restic version`

Any.

## Expected behavior

Adding an extra column to list the size of the backup (in bytes) can be very useful.
It'll help distinguish between different backups just by checking  their size.

```
$ restic snapshots
ID        Date                 Host        Tags        Directory    Size
--------------------------------------------------------------------------
5b969a0e  2016-12-09 15:10:32  localhost               myfile       390865
```

## Actual behavior

```
$ restic snapshots
ID        Date                 Host        Tags        Directory
----------------------------------------------------------------------
5b969a0e  2016-12-09 15:10:32  localhost               myfile
```

### LINKED ISSUE #874: More stats in snapshots list
It would be useful to have a more stats that are not currently in snapshots list like:
- Backup size
- Backup duration or end date

Thanks :)

### PR DESCRIPTION
**Pull request by MichaelEischer**
What does this PR change? What problem does it solve?
-----------------------------------------------------
Store the backup statistics in a snapshot. Sample output:
```
{
  "time": "2024-02-22T22:39:19.991058932+01:00",
  "parent": "2c29113d80ef2f4072bae3095963a9c70f9d7014e893afd74887c8b61d30f143",
  "tree": "c163608e60d5bf1921b296528cd28b9844b508d96d9f4ad2564ced61c6dd0a24",
  "paths": [
    "/home/michael/Projekte/restic/restic"
  ],
  "hostname": "host",
  "username": "michael",
  "uid": 1000,
  "gid": 1000,
  "program_version": "restic 0.16.4-dev (compiled manually)",
  "summary": {
    "backup_start": "2024-02-22T22:39:19.991058932+01:00",
    "backup_end": "2024-02-22T22:39:26.37292521+01:00",
    "files_new": 85,
    "files_changed": 155,
    "files_unmodified": 2493,
    "dirs_new": 3,
    "dirs_changed": 110,
    "dirs_unmodified": 328,
    "data_blobs": 234,
    "tree_blobs": 114,
    "data_added": 39375231,
    "data_added_in_repo": 19236421,
    "total_files_processed": 2733,
    "total_bytes_processed": 152165603
  }
}
```

In addition the `snapshots` command now also prints the snapshot size (based on `total_bytes_processed`).

```
ID        Time                 Host        Tags        Paths                                    Size
-----------------------------------------------------------------------------------------------------------
4e0ff0ef  2024-02-22 22:39:19  host                    /home/michael/Projekte/restic/restic     145.116 MiB
-----------------------------------------------------------------------------------------------------------
```

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
Fixes https://github.com/restic/restic/issues/693
Fixes https://github.com/restic/restic/issues/874

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

**Comment by darkdragon-001** [comment]
Does the `--json` output also contain the size?

**Comment by aawsome** [comment]
@MichaelEischer Is there a reason why your implementation deviates from
https://github.com/restic/restic/issues/693#issuecomment-1145295573 ?

**Comment by aawsome** [comment]
> Differences to [#693 (comment)](https://github.com/restic/restic/issues/693#issuecomment-1145295573):
> 
>     * "data_added_in_repo" instead of "data_added_packed": the corresponding fields in `ItemStats` are called `DataSizeInRepo` and `TreeSizeInRepo`. Thus that name is more consistent.

If you name them like this you make the snapshot format incompatible to what rustic implemented some 2 years ago. I don't care too much if you omit the other fields, but same things should be named the same, IMO.

**Comment by hraban** [comment]
Just for the record this is only meaningful until you start deleting other snapshots which share data, right?

**Comment by aawsome** [comment]
> Just for the record this is only meaningful until you start deleting other snapshots which share data, right?

No. It is as meaningful as it is saving the output of `backup --json` in some database. If you are trying to derive "how much size does this snapshot uniquely occupy in the repo" from the information you are making a mistake as this is not what it states. This information cannot be derived even after a follow-up backup run (which may share data).
The information is only a "snapshot" of what happened during the backup run. Including information about the source (which could also change directly after the backup run). But for analysis reasons the information might be interesting.

**Comment by rawtaz** [comment]
Regardless of other factors and what other software projects have done, restic developers should naturally be free to name things in ways they think makes sense and aligns with restic and its code base.

**Comment by aawsome** [comment]
@rawtaz If this is the way the restic project thinks about free open source software community...

**Comment by MichaelEischer** [comment]
I've renamed the attribute to `data_added_packed`. The text output uses that value in `%s stored` which might also have served as a suffix. But that's just bikeshedding, I'm fine with either variant. So let's use the one that maintains compatibility.

In general, we're interested in a friendly coexistence with rustic. But it should also be clear that we won't always agree how things should be implemented. Nevertheless there's still value in trying to maintain compatibility where it makes sense. That judgement will still be up to us, but that doesn't mean the we do X just because rustic implemented Y.

> Is there a reason why your implementation deviates from #693 (comment) ?

Let me add just one thing here: the PR was intentionally still marked as draft as the naming issue was not yet resolved.

**Comment by MichaelEischer** [comment]
_(separate comment as I intend to mark the last three comments of this argument as offtopic)_

@aawsome @rawtaz Stop going at each other's throats, that applies to both of you!

Mostly a reply to @aawsome:

> Regardless of other factors and what other software projects have done, restic developers should naturally be free to name things in ways they think makes sense and aligns with restic and its code base.

Without context that's a true statement, although it doesn't mean that established conventions have to be ignored, just that we are free to do so if deemed beneficial. But I see how it can be read as "we're not interested in your input" in the context of this PR. For my take on that see the coexistence paragraph above.

> If this is the way the restic project thinks about free open source software community...

That's an improper generalization. rustic is not equivalent to the "community" (to prevent misinterpretations: rustic is obviously part of the community).

One more thing: while I understand that you're disappointed about how this discussion went, just closing all your PRs (which were still open as I've said several times that we're still interested in most of those changes) over an argument about a single attribute name is IMHO really excessive. I still intend to pick up useful parts from them.

**Comment by rawtaz** [comment]
> Stop going at each other's throats, that applies to both of you!

Yessir! 🫡

**Comment by aawsome** [comment]
> I've renamed the attribute to `data_added_packed`.

Thanks for changing this. I think this helps user who use both restic and rustic. And it's about our users what we are doing, right?
I will change rustic such that it can work with omitting the fields you don't fill in restic.

> In general, we're interested in a friendly coexistence with rustic.

Good to hear. I am also very interested in a friendly coexistence. This is all free open source software!

**Comment by aawsome** [comment]
(as this all goes off-topic, I answer here)

> @aawsome @rawtaz Stop going at each other's throats, that applies to both of you!

I don't ever intended or intend to go at one other's throat.

> Mostly a reply to @aawsome:
> 
> > Regardless of other factors and what other software projects have done, restic developers should naturally be free to name things in ways they think makes sense and aligns with restic and its code base.
> 
> Without context that's a true statement, although it doesn't mean that established conventions have to be ignored, just that we are free to do so if deemed beneficial. But I see how it can be read as "we're not interested in your input" in the context of this PR. For my take on that see the coexistence paragraph above.

Yes, it's about context.

> > If this is the way the restic project thinks about free open source software community...
> 
> That's an improper generalization. rustic is not equivalent to the "community" (to prevent misinterpretations: rustic is obviously part of the community).

This was the answer to the comment which was about "Regardless of other factors and what other software projects have done". 

> One more thing: while I understand that you're disappointed about how this discussion went, just closing all your PRs (which were still open as I've said several times that we're still interested in most of those changes) over an argument about a single attribute name is IMHO really excessive. I still intend to pick up useful parts from them.

About PRs: I am no longer willing to contribute to restic in future. This is a personal decision and I hope you respect it. There are various very personal reasons for it:
- I am pretty short on time and have to take priorities
- I don't like to program in Go in my free time as Rust pleases me much more
- All features I wanted or want in restic I did or can implement in rustic
- I was very much disappointed how contributing to restic went, mainly the very long times until things were processed and some endless discussions about "do we really need this feature?" - I feel I have another mentality here
- Finally, the feeling of hostility here and in the forum (and I don't mean that people get reasonably annoyed when I wrote 3 times that rustic supports a feature where 1 time might have sufficed)
- The discussion on this PR is not the reason for doing anything but might have contributed to the last part.

You as restic project may (but don't have to) think about whether these are things you are satisfied with or which you might want to change.
I myself am trying to do it another way with rustic, but I don't know if I am successful. And, BTW: I warmly invite you to join rustic. Programming in Rust is fun and worth a consideration for every developer!

Now back to the PRs: Given that everything I did was more than 2 years old, I don't have too much track of what happened in restic since, and I am not willing to contribute further, I decided to remove those together with other stone-old stuff.
All I can say about them is that everything important is implemented in rustic and I would have to look into the rustic codebase or try to understand what I wrote 2 years ago. I don't see that there is value in it. If you still have some code snippets, use what you want. But most likely it would be easier to re-implement and (if needed) take the rustic implementation as reference. Of course, I offer to support you on conceptual questions. But IIRC those things were not rocket-science, but merely "just do it"-things.

**Review by MichaelEischer — COMMENTED** [review]
LGTM.

File: changelog/unreleased/issue-693 [hunk 1]
```diff
@@ -0,0 +1,12 @@
+Enhancement: Support printing snapshot size in `snapshots` command
+
+The `snapshots` command now supports printing the snapshot size for snapshots
+created using this or a future restic version. For this, the `backup` command
+now stores the backup summary statistics in the snapshot.
+
+The text output of the `snapshots` command only shows the snapshot size. The
```
Comment (mirabilos) [inline changelog/unreleased/issue-693:7]: What exactly of all the discussed options is the “snapshot size” here?

Bytes added to the repo when this backup was created?

If so, fine with me, but I would _still_ **love** to also see the “amount of data backed up” (what others call restore size) for a quick double-check.

Comment (aawsome) [reply changelog/unreleased/issue-693:7]: It is `total_bytes_processed` which IMO should be the restore size.

**Pull request merged**